lib/solarwinds_apm/support/resource_detector.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+# © 2023 SolarWinds Worldwide, LLC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at:http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+require 'net/http'
+require 'uri'
+require 'json'
+
+module SolarWindsAPM
+  # ResourceDetector
+  # Usage:
+  #   require 'opentelemetry/sdk'
+  #   require 'opentelemetry/resource/detector'
+  #   OpenTelemetry::SDK.configure do |c|
+  #     c.resource = SolarWindsAPM::ResourceDetector.detect
+  #   end
+  module ResourceDetector
+    K8S_PODNAME_PATH = '/etc/hostname'
+    K8S_NAMESPACE_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
+    K8S_TOKEN_PATH = '/var/run/secrets/kubernetes.io/serviceaccount/token'
+
+    UAMS_CLIENT_PATH = '/opt/solarwinds/uamsclient/var/uamsclientid'
+    UAMS_CLIENT_URL = 'http://127.0.0.1:2113/info/uamsclient'
+    UAMS_CLIENT_ID_FIELD = 'uamsclient_id'
+
+    def self.detect
+      attributes = ::OpenTelemetry::SDK::Resources::Resource.create({})
+      attributes = attributes.merge(detect_uams_client_id)
+      attributes = attributes.merge(detect_k8s_atttributes)
+      attributes.merge(from_upstream_detector)
+    end
+
+    def self.detect_uams_client_id
+      uams_client_id = nil
+      if File.exist?(UAMS_CLIENT_PATH)
+        uams_client_id = File.read(UAMS_CLIENT_PATH).strip
+      else
+        url = URI(UAMS_CLIENT_URL)
+
+        response = nil
+        ::OpenTelemetry::Common::Utilities.untraced do
+          http = Net::HTTP.new(url.host, url.port)
+          request = Net::HTTP::Get.new(url)
+          response = http.request(request)
+        end
+
+        if response&.code.to_i == 200
+          uams_metadata = JSON.parse(response.body)
+          uams_client_id = uams_metadata&.fetch(UAMS_CLIENT_ID_FIELD)
+        end
+      end
+
+      resource_attributes = {
+        'sw.uams.client.id' => uams_client_id
```

Comment:
Our spec requires this to be set as the `host.id` attribute.

Reply:
Just to confirm that `host.id` == `sw.uams.client.id`?

Reply:
Yes. you can leave the setting of `sw.uams.client.id` attribute if you'd like, should do no harm. But for entity inference to work this value needs to be set as the `host.id` attribute, see the "SWO proprietary logic for the `host.id` attribute" section in https://github.com/solarwinds-cloud/trace/blob/master/docs/specs/resource-attributes.md#host